### PR TITLE
[SUP-802] Added logic to prevent a disk from being deleted when it's in a valid state

### PR DESCRIPTION
**Pull request by n8landolt**
https://broadworkbench.atlassian.net/browse/SUP-802

This ticket fixes a bug where a runtime errors out and deletes the user's PD without request.

To recreate - make a jupyter runtime, delete the runtime, create a new runtime with a failing user script, and then the PD is set to deleted, however it still exists on the Google side.

For reviewers - is this added logic the outcome we want? Or do we want to just leave the PD alone? "Acceptance critera" isn't defined, so I made some assumptions about the outcome.

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green (you can re-run automation tests with a comment saying `jenkins retest`
- [ ] Run the automation tests multiple times in parallel to weed out instability if applicable via a comment saying `multi-test`
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala [hunk 1]
```diff
@@ -167,11 +167,17 @@ abstract class BaseCloudServiceRuntimeMonitor[F[_]] {
...
                   persistentDiskQuery.getPersistentDiskRecord(did).transaction
                 )
-                _ <- persistentDiskOpt.traverse_(d =>
-                  googleDisk.deleteDisk(googleProject, rc.zone, d.name) >> persistentDiskQuery
-                    .updateStatus(d.id, DiskStatus.Deleted, ctx.now)
-                    .transaction
-                )
+                _ = persistentDiskOpt match {
+                  case Some(value) =>
+                    if (value.status == DiskStatus.Creating || value.status == DiskStatus.Failed) {
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala:172]: Am I understanding correctly that we want to delete any disk in a non ready status? In that case, it might be easier to have the following condition:
`if (value.status != DiskStatus.Ready)`

Comment (LizBaldo) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala:172]: Regardless, adding a comment explaining the expected behavior would be nice :)

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitorSpec.scala [hunk 2]
```diff
@@ -255,7 +255,7 @@ class BaseCloudServiceRuntimeMonitorSpec extends AnyFlatSpec with Matchers with
       // handleCheckTools should have timed out after 10 seconds and the runtime should remain in Deleted status
       elapsed should be >= 10000L
       status shouldBe Some(RuntimeStatus.Deleted)
-      diskStatus shouldBe (Some(DiskStatus.Deleted))
+      diskStatus shouldBe (Some(DiskStatus.Ready))
```
Comment (LizBaldo) [inline http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitorSpec.scala:258]: For this file and the next, it would be nice to keep the existing test that triggers the disk deletion)

**Review by LizBaldo — COMMENTED** [review]
Am I understanding correctly that

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala (hunk 1 above)
Comment (n8landolt) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala:172]: Personally, I don't think we should affect the PD at all. But if we do want to, do we know what the 'Restoring' status is used for? I would think we don't want to delete a 'Restoring' disk as well.

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitorSpec.scala (hunk 2 above)
Comment (n8landolt) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitorSpec.scala:258]: Agreed! I'll get a test in there before I request your review again.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala (hunk 1 above)
Comment (jdcanas) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala:172]: I like the idea of being explicit in the statuses we want to delete instead of the inverse. Because users often save disks between runtimes, I'd hate to eagerly delete.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala [hunk 3]
```diff
@@ -1365,11 +1364,17 @@ class LeoPubsubMessageSubscriber[F[_]](
...
                         persistentDiskQuery.getPersistentDiskRecord(did).transaction
                       )
-                      _ <- persistentDiskOpt.traverse_(d =>
-                        googleDiskService.deleteDisk(googleProject, rc.zone, d.name) >> persistentDiskQuery
-                          .updateStatus(d.id, DiskStatus.Deleted, now)
-                          .transaction
-                      )
+                      _ = persistentDiskOpt match {
+                        case Some(value) =>
+                          if (value.status == DiskStatus.Creating || value.status == DiskStatus.Failed) {
```
Comment (jdcanas) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala:1369]: Huh... I'm surprised this compiles?

This is probably the reason your test is failing. You never actually run the IO from `persistentDiskOpt.traverse_(d =>
                              googleDiskService.deleteDisk(googleProject, rc.zone, d.name) >> persistentDiskQuery
                                .updateStatus(d.id, DiskStatus.Deleted, now)
                                .transaction
                            )`

In other words, Line 1367 should read 
`_ <- persistentDiskOpt match {` not `_ = persistentDiskOpt match {`

Furthermore, that will cause a syntax error, as the type from the `if() { ... }` branch is F[_] and the type of the else branch is None (aka not included, and should likely be `else F.unit`)

**Comment by mlilynolting** [comment]
How would I test the two distinct error handlers, as a reviewer? The LeoPubSubMessageSubscriber and the BaseCloudServiceRuntimeMonitor presumably are triggering these sections of code in different situations, and I'm not sure when one would fire over the other.

**Review by LizBaldo — APPROVED** [review]
Thanks for putting back the tests!

You have my 👍, assuming that you confirmed the logic and expected user experience with Alessandro?

**Review by jdcanas — APPROVED** [review]

**Comment by n8landolt** [comment]
jenkins retest

**Comment by n8landolt** [comment]
jenkins retest

**Comment by n8landolt** [comment]
jenkins retest
